File: implementation/scipamato/core/web/src/main/java/ch/difty/scipamato/core/config/ScipamatoCoreProperties.java
package ch.difty.scipamato.core.config;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.stereotype.Component;

import ch.difty.scipamato.common.config.AbstractScipamatoProperties;
import ch.difty.scipamato.common.config.MavenProperties;
import ch.difty.scipamato.core.logic.exporting.RisExporterStrategy;
import ch.difty.scipamato.core.logic.parsing.AuthorParserStrategy;

/**
 * This bean is used to evaluate all environment properties used in the
 * application in one place and serve those as bean to wherever they are used.
 *
 * @author u.joss
 * @see <a href=
 *     "https://www.petrikainulainen.net/programming/spring-framework/spring-from-the-trenches-injecting-property-values-into-configuration-beans/">https://www.petrikainulainen.net/programming/spring-framework/spring-from-the-trenches-injecting-property-values-into-configuration-beans/</a>
 */
@Component
public class ScipamatoCoreProperties extends AbstractScipamatoProperties<ScipamatoProperties>
    implements ApplicationCoreProperties {

    private static final long serialVersionUID = 1L;

    public ScipamatoCoreProperties(@NotNull final ScipamatoProperties scipamatoProperties,
        @NotNull final MavenProperties mavenProperties) {
        super(scipamatoProperties, mavenProperties);
    }

    @NotNull
    @Override
    public AuthorParserStrategy getAuthorParserStrategy() {
        return getScipamatoProperties().getAuthorParserStrategy();
    }

    @NotNull
    @Override
    public RisExporterStrategy getRisExporterStrategy() {
        return getScipamatoProperties().getRisExporterStrategy();
    }

    @Override
    public long getMinimumPaperNumberToBeRecycled() {
        return getScipamatoProperties().getPaperNumberMinimumToBeRecycled();
    }

    @Nullable
    @Override
    public String getPubmedApiKey() {
        return getScipamatoProperties().getPubmedApiKey();
    }
}

File: core/core-web/src/main/java/ch/difty/scipamato/core/config/ScipamatoCoreProperties.kt
@file:Suppress("MaxLineLength")

package ch.difty.scipamato.core.config

import ch.difty.scipamato.common.config.AbstractScipamatoProperties
import ch.difty.scipamato.common.config.MavenProperties
import ch.difty.scipamato.core.logic.exporting.RisExporterStrategy
import ch.difty.scipamato.core.logic.parsing.AuthorParserStrategy
import org.springframework.stereotype.Component

/**
 * This bean is used to evaluate all environment properties used in the
 * application in one place and serve those as bean to wherever they are used.
 *
 * @see [https://www.petrikainulainen.net/programming/spring-framework/spring-from-the-trenches-injecting-property-values-into-configuration-beans/](https://www.petrikainulainen.net/programming/spring-framework/spring-from-the-trenches-injecting-property-values-into-configuration-beans/)
 */
@Component
class ScipamatoCoreProperties(
    scipamatoProperties: ScipamatoProperties,
    mavenProperties: MavenProperties,
) : AbstractScipamatoProperties<ScipamatoProperties>(
    scipamatoProperties,
    mavenProperties
), ApplicationCoreProperties {

    override val authorParserStrategy: AuthorParserStrategy
        get() = scipamatoProperties.authorParserStrategy
    override val risExporterStrategy: RisExporterStrategy
        get() = scipamatoProperties.risExporterStrategy
    override val minimumPaperNumberToBeRecycled: Long
        get() = scipamatoProperties.paperNumberMinimumToBeRecycled.toLong()
    override val pubmedApiKey: String?
        get() = scipamatoProperties.pubmedApiKey

    companion object {
        private const val serialVersionUID = 1L
    }
}
